ui/library/src/context/PageRenderContext.ts

```diff
@@ -185,6 +188,7 @@ async function buildPageObjectURL({
     const viewport = pageProxy.getViewport({ scale: scale * zoomMultiplier * devicePixelRatio });
     canvas.height = viewport.height;
     canvas.width = viewport.width;
+    canvas.style.width = '100%'; //de-zoom canvas with style (maybe you can directly use CSS), reaching de-zoom of higher definition PDF
```

Comment:
Style is meaningless when the element isn't in the DOM.

Reply:
You _might_ try `canvasContext.imageSmoothingEnabled = false;`.

Another thing you could try is `canvasContext.imageSmoothingQuality = 'high';`.